src/main/java/carpet/script/api/WorldAccess.java

```diff
@@ -1627,23 +1631,69 @@ else if (ticket == TicketType.POST_TELEPORT) // post teleport
-            ret.put(new StringValue(noise), new NumericValue(noiseValue));
+        Map<Value, Value> result = new HashMap<>();
+
+        ServerLevel level = ((CarpetContext) c).level();
+        ChunkGenerator chunkGenerator = level.getChunkSource().getGenerator();
+
+        if (chunkGenerator instanceof NoiseBasedChunkGenerator generator) {
+
+            var randomState = RandomState
+                    .create(generator.generatorSettings().value(), level.registryAccess()
+                    .lookupOrThrow(Registries.NOISE), level.getSeed());
+
+            var densityFunctionRegistry = level.registryAccess().registryOrThrow(Registries.DENSITY_FUNCTION);
+
+            var visitor = ((RandomStateVisitorAccessor) (Object) randomState).getVisitor();
+
+            var singlePointContext = new DensityFunction.SinglePointContext(pos.getX(), pos.getY(), pos.getZ());
+            var router = generator.generatorSettings().value().noiseRouter();
+
+            for (String densityFunctionQuery : densityFunctionQueries) {
+                var densityFunction = switch (densityFunctionQuery) {
+                    case "barrier_noise" -> router.barrierNoise();
+                    case "fluid_level_floodedness_noise" -> router.fluidLevelFloodednessNoise();
+                    case "fluid_level_spread_noise" -> router.fluidLevelSpreadNoise();
+                    case "lava_noise" -> router.lavaNoise();
+                    case "temperature" -> router.temperature();
+                    case "vegetation" -> router.vegetation();
+                    case "continents" -> router.continents();
+                    case "erosion" -> router.erosion();
+                    case "depth" -> router.depth();
+                    case "ridges" -> router.ridges();
+                    case "initial_density_without_jaggedness" -> router.initialDensityWithoutJaggedness();
+                    case "final_density" -> router.finalDensity();
+                    case "vein_toggle" -> router.veinToggle();
+                    case "vein_ridged" -> router.veinRidged();
+                    case "vein_gap" -> router.veinGap();
+                    default -> null;
+                };
+
+                if (densityFunction == null) {
+                    ResourceLocation densityFunctionKey = ResourceLocation.tryParse(densityFunctionQuery);
+                    if (densityFunctionKey == null) {
+                        throw new InternalExpressionException("Density function name '" + densityFunctionQuery + "' is invalid. Please follow the correct [a-z0-9._-]:[a-z0-9/._-] format.");
+                    }
+
+                    if (densityFunctionRegistry.containsKey(densityFunctionKey)) {
+                        densityFunctionQuery = densityFunctionKey.toString();
+                        densityFunction = densityFunctionRegistry.get(densityFunctionKey);
```

Comment:
Given the `densityFunctionQuery` is used for mapping the results, why change it here? That means it's possible these won't be the same and it may be confusing. Also I believe ResLoc.toString returns with the `minecraft` namespace, while scarpet usually skips it (see `ValueConversions`)